src/main/java/me/zeroeightsix/fiber/tree/IndexedNodeCollection.java

```diff
@@ -0,0 +1,116 @@
+package me.zeroeightsix.fiber.tree;
+
+import me.zeroeightsix.fiber.exception.DuplicateChildException;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.*;
+
+public class IndexedNodeCollection extends AbstractCollection<ConfigNode> implements NodeCollection {
+    private final Map<String, ConfigNodeImpl> items = new TreeMap<>();
+    @Nullable private final ConfigBranch owner;
+
+    public IndexedNodeCollection(@Nullable ConfigBranch owner) {
+        this.owner = owner;
+    }
+
+    public IndexedNodeCollection(@Nullable ConfigBranch owner, Collection<ConfigNode> items) {
+        this(owner);
+        this.addAll(items);
+    }
+
+    @Nonnull
+    @Override
+    public Iterator<ConfigNode> iterator() {
+    return new Iterator<ConfigNode>() {
+            @Nullable private ConfigNodeImpl last;
+            private Iterator<ConfigNodeImpl> backing = items.values().iterator();
+
+            @Override
+            public boolean hasNext() {
+                return backing.hasNext();
+            }
+
+            @Override
+            public ConfigNode next() {
+                this.last = this.backing.next();
+                return last;
+            }
+
+            @Override
+            public void remove() {
+                if (this.last == null) throw new IllegalStateException();
+                this.backing.remove();
+                this.last.detach();
+            }
+        };
+    }
+
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    @Override
+    public Spliterator<ConfigNode> spliterator() {
+        // The Spliterator interface is read-only, so we consider its element type covariant
+        return (Spliterator/*<out ConfigNode>*/) this.items.values().spliterator();
+    }
+
+    @Override
+    public boolean add(ConfigNode item) throws DuplicateChildException {
+        return add(item, false);
+    }
+
+    @Override
+    public boolean add(ConfigNode item, boolean overwrite) throws DuplicateChildException {
+        Objects.requireNonNull(item);
+        if (!(item instanceof ConfigNodeImpl)) {
+            throw new IllegalArgumentException("Invalid config node implementation " + item);
+        }
+        if (overwrite) {
+            this.removeByName(item.getName());
+        } else if (this.items.containsKey(item.getName())) {
+            throw new DuplicateChildException("Attempt to replace node " + item.getName());
+        }
+        this.items.put(item.getName(), (ConfigNodeImpl) item);
+        ((ConfigNodeImpl) item).setParent(this.owner);
+        return true;
```

Comment:
Why must `item` be an instance of `ConfigNodeImpl`? Should we be preventing people from implementing `ConfigNode`?

If needed, just add the `setParent` method to `ConfigNode`.

Reply:
Alright, we can probably expose those with a few adjustments.